#2663 Kafka Streams Extension Guide (#2896)
quarkusio/quarkus: docs/src/main/asciidoc/kafka-streams-guide.adoc

Another topic (`weather-stations`) contains just the master data about the weather stations themselves (id and name).

A second component (`aggregator`) reads from the two Kafka topics and processes them in a streaming pipeline:

* the two topics are joined on weather station id
* per weather station the min, max and average temperature is determined
* this aggregated data is written out to a third topic (`temperatures-aggregated`)

The data can be examined by inspecting the output topic.
By exposing a Kafka Streams https://kafka.apache.org/22/documentation/streams/developer-guide/interactive-queries.html[interactive query],
the latest result for each weather station can alternatively be obtained via a simple REST query.

The overall architecture looks like so:

image::kafka-streams-guide-architecture.png[alt=Architecture]

== Solution

We recommend that you follow the instructions in the next sections and create the application step by step.
However, you can go right to the completed example.

Clone the Git repository: `git clone {quickstarts-clone-url}`, or download an {quickstarts-archive-url}[archive].

> The solution is located in the `kafka-quickstart` {quickstarts-tree-url}/kafka-streams-quickstart[directory].

Reviewer: ```suggestion
The solution is located in the `kafka-streams-quickstart` {quickstarts-tree-url}/kafka-streams-quickstart[directory].
```

 */
@ApplicationScoped
public class ValuesGenerator {

    private static final Logger LOG = LoggerFactory.getLogger(ValuesGenerator.class);

    private Random random = new Random();

    private List<WeatherStation> stations = Collections.unmodifiableList(
            Arrays.asList(
                    new WeatherStation(1, "Hamburg", 13),
                    new WeatherStation(2, "Snowdonia", 5),
                    new WeatherStation(3, "Boston", 11),
                    new WeatherStation(4, "Tokio", 16),
                    new WeatherStation(5, "Cusco", 12),
                    new WeatherStation(6, "Svalbard", -7),
                    new WeatherStation(7, "Porthsmouth", 11),
                    new WeatherStation(8, "Oslo", 7),
                    new WeatherStation(9, "Marrakesh", 20)
            ));

    @Outgoing("temperature-values")                             // <1>
    public Flowable<KafkaMessage<Integer, String>> generate() {

        return Flowable.interval(500, TimeUnit.MILLISECONDS)    // <2>
                .map(tick -> {
                    WeatherStation station = stations.get(random.nextInt(stations.size()));

>                     double temperature = new BigDecimal(random.nextGaussian() * 15 + station.averageTemperature)

Reviewer: The width of the line is a bit too long for the web site.

Author: Wrapped it.

                });
    }

    @Outgoing("weather-stations")                               // <3>
    public Flowable<KafkaMessage<Integer, String>> weatherStations() {
        List<KafkaMessage<Integer, String>> stationsAsJson = stations.stream()
            .map(s -> KafkaMessage.of(s.id, "{ \"id\" : " + s.id + ", \"name\" : \"" + s.name + "\" }"))
            .collect(Collectors.toList());

        return Flowable.fromIterable(stationsAsJson);
    };

    private static class WeatherStation {

        int id;
        String name;
        int averageTemperature;

        public WeatherStation(int id, String name, int averageTemperature) {
            this.id = id;
            this.name = name;
            this.averageTemperature = averageTemperature;
        }
    }
}
----
<1> Instruct Reactive Messaging to dispatch the items from the returned `Flowable` to `temperature-values`.
<2> The method returns a RX Java 2 _stream_ (`Flowable`) emitting a random temperature value every 0.5 seconds.
<3> Instruct Reactive Messaging to dispatch the items from the returned `Flowable` (static list of weather stations) to `weather-stations`.

> The two method each return a _reactive stream_ whose items are sent to the streams named `temperature-values` and `weather-stations`, respectively.

Reviewer: ```suggestion
The two methods each return a _reactive stream_ whose items are sent to the streams named `temperature-values` and `weather-stations`, respectively.
```

=== Topic Configuration

> The two streams are mapped to Kafka Apache topics using the Quarkus configuration file `application.properties`.

Reviewer: ```suggestion
The two channels are mapped to Kafka topics using the Quarkus configuration file `application.properties`.
```

For that, add the following to the file `producer/src/main/resources/application.properties`:

[source]
----

> kafka.bootstrap.servers=localhost:9092

Reviewer: Maybe you can add a comment:
# Configure the Kafka broker location

Author: Done.

mp.messaging.outgoing.temperature-values.connector=smallrye-kafka
mp.messaging.outgoing.temperature-values.key.serializer=org.apache.kafka.common.serialization.IntegerSerializer
mp.messaging.outgoing.temperature-values.value.serializer=org.apache.kafka.common.serialization.StringSerializer

mp.messaging.outgoing.weather-stations.connector=smallrye-kafka
mp.messaging.outgoing.weather-stations.key.serializer=org.apache.kafka.common.serialization.IntegerSerializer
mp.messaging.outgoing.weather-stations.value.serializer=org.apache.kafka.common.serialization.StringSerializer
----

This configures the Kafka bootstrap server, the two topics and the corresponding (de-)serializers.

> More details about the different configuration options are available on the https://kafka.apache.org/documentation/#producerconfigs[Producer configuration] and https://kafka.apache.org/documentation/#consumerconfigs[Consumer configuration] section from the Kafka documentation.

Reviewer: Not really a comment, but do you know another place or a cheatsheet with the same content?

Author: No, I think that's the authoritative source.

== Creating the Aggregator Maven Project

With the producer application in place, it's time to implement the actual aggregator application,
which will run the Kafka Streams pipeline.
Create another project like so:

[source, subs=attributes+]
----
mkdir aggregator && cd aggregator && \
mvn io.quarkus:quarkus-maven-plugin:{quarkus-version}:create \
    -DprojectGroupId=org.acme \
    -DprojectArtifactId=kafka-streams-quickstart-aggregator \
    -Dextensions="kafka-streams,resteasy-jsonb" \
    && cd ..
----

This creates the `aggregator, project with the Quarkus extension for Kafka Streams and with RESTEasy support for JSON-B.

=== The Pipeline Implementation

Let's begin the implementation of the stream processing application by creating
a few value objects for representing temperature measurements, weather stations and for keeping track of aggregated values.

First, create the file `aggregator/src/main/java/org/acme/quarkus/sample/kafkastreams/model/WeatherStation.java`,
representing a weather station, with the following content:

[source, java]
----
package org.acme.quarkus.sample.kafkastreams.model;

import io.quarkus.runtime.annotations.RegisterForReflection;

> @RegisterForReflection

Reviewer: I have the same issue with Reactive Messaging. We need to look into that and detect this case in the extension.

Author: Yes, for Kafka Streams it'd probably make sense to auto-register all those types which are de-serialized via a `Serde` used in the pipeline.

Author: Added this as one task to https://github.com/quarkusio/quarkus/issues/2863.

Reviewer: Thanks!

public class WeatherStation {

    public int id;
    public String name;
}
----

Then the file `aggregator/src/main/java/org/acme/quarkus/sample/kafkastreams/model/TemperatureMeasurement.java`,
representing temperature measurements for a given station:

[source, java]
----
package org.acme.quarkus.sample.kafkastreams.model;

import java.time.Instant;

> public class TemperatureMeasurement {

Reviewer: Don't you need the `@RegisterForReflection`

Author: No, it's only instantiated once in `KafkaStreamsPipeline`, calling the constructor.

        this.value = value;
    }
}
----

And finally `aggregator/src/main/java/org/acme/quarkus/sample/kafkastreams/model/Aggregation.java`,
which will be used to keep track of the aggregated values while the events are processed in the streaming pipeline:

[source, java]
----
package org.acme.quarkus.sample.kafkastreams.model;

import java.math.BigDecimal;
import java.math.RoundingMode;

import io.quarkus.runtime.annotations.RegisterForReflection;

@RegisterForReflection
public class Aggregation {

    public int stationId;
    public String stationName;
    public double min = Double.MAX_VALUE;
    public double max = Double.MIN_VALUE;
    public int count;
    public double sum;
    public double avg;

    public Aggregation updateFrom(TemperatureMeasurement measurement) {
        stationId = measurement.stationId;
        stationName = measurement.stationName;

        count++;
        sum += measurement.value;
        avg = BigDecimal.valueOf(sum / count)
                .setScale(1, RoundingMode.HALF_UP).doubleValue();

        min = Math.min(min, measurement.value);
        max = Math.max(max, measurement.value);

        return this;
    }
}
----

> All these types carry the `@RegisterForReflection`, allowing them to be instantiated reflectively when running the application in native mode.

Reviewer: Except one :-)

Reviewer: Actually, the reason is to instruct the Graal compiler to make sure the classes are not removed during the dead-code elimination

Author: The classes are referenced in the code, so they shouldn't be removed entirely. It's only the parameterless constructor that'll be removed. In fact, just re-tested in native mode, and it's only needed for `WeatherStation`, as that's the only type instantiated reflectively by a `Deserializer`. Pushed a commit for removing the superfluous ones and moving the sentence right to that listing itself.

    ]
    depends_on:
      - zookeeper
    ports:
      - "9092:9092"
    environment:
      LOG_DIR: "/tmp/logs"
      KAFKA_ADVERTISED_LISTENERS: PLAINTEXT://kafka:9092
      KAFKA_LISTENERS: PLAINTEXT://0.0.0.0:9092
      KAFKA_ZOOKEEPER_CONNECT: zookeeper:2181
      KAFKA_NUM_PARTITIONS: 3
    networks:
      - kafkastreams-network

  producer:
    image: quarkus-quickstarts/kafka-streams-producer:1.0
    build:
      context: producer
      dockerfile: src/main/docker/Dockerfile.${QUARKUS_MODE:-jvm}
    environment:
      KAFKA_BOOTSTRAP_SERVERS: kafka:9092
    networks:
      - kafkastreams-network

  aggregator:
    image: quarkus-quickstarts/kafka-streams-aggregator:1.0
    build:
      context: aggregator
      dockerfile: src/main/docker/Dockerfile.${QUARKUS_MODE:-jvm}
    environment:
      ORG_ACME_QUARKUS_SAMPLE_KAFKASTREAMS_BOOTSTRAP_SERVERS: kafka:9092
    networks:
      - kafkastreams-network

networks:
  kafkastreams-network:
    name: ks
----

To launch all the containers, building the `producer` and `aggregator` container images,

> run `docker-compose up --build`.

Reviewer 2: Should the command only be `docker-compose up` which will build the image when necessary as opposed to `docker-compose up --build` which will build the image each time the command is launched ?

Author: I think it's better for these instructions to rather build once too often than forgetting it once. E.g. if you re-do them a second time, you might easily use the old images unintentionally.

You should see log statements from the `producer` application about messages being sent to the "temperature-values" topic.

> Due to an [an issue](https://github.com/smallrye/smallrye-reactive-messaging/issues/128) in the SmallRye Reactive Messaging component it might be that the _producer_ application isn't sending events if it was started before Apache Kafka was up.

Reviewer: It has been fixed in the latest released. It will restart sending messages when Kafka is available. However, if the source of the events does not have back pressure, it may fail there. So you need a `onBackPRessureDrop()`

Author: Where does that go? Do you have an example handy?

Author: Added it. Thanks!
